ext/spl/spl_array.c

```diff
@@ -163,11 +161,11 @@ static zend_object *spl_array_object_new_ex(zend_class_entry *class_type, zend_o
 		if (clone_orig) {
 			if (other->ar_flags & SPL_ARRAY_IS_SELF) {
 				ZVAL_UNDEF(&intern->array);
-			} else if (orig->handlers == &spl_handler_ArrayObject) {
+			} else if (instanceof_function(class_type, spl_ce_ArrayObject)) {
 				ZVAL_ARR(&intern->array,
 					zend_array_dup(spl_array_get_hash_table(other)));
 			} else {
-				ZEND_ASSERT(orig->handlers == &spl_handler_ArrayIterator);
+				ZEND_ASSERT(instanceof_function(class_type, spl_ce_ArrayIterator));
```

Comment:
Even though asserts are gone in release builds, the call to `instanceof_function` will remain because the compiler can't prove in this compile unit that this function is side-effect-free. It's a bit unfortunate.

Reply:
Ah, this is unfortunate indeed, would adding `__attribute__((const))` (GCC) attributes to `instanceof_function` and `instanceof_function_slow` fix this issue?